### PR TITLE
Add workaround to fix system's migration tasks for 2.6 to 2.7

### PR DESCRIPTION
**Pull request by miguelsorianod**

### CODE REVIEW

**Comment by miguelsorianod** [comment]
Tried deployment with the new command when deploying from zero.
Now at the end of the pre-hook pod I observe the following logs:

```
Migrating proxies api_backend to backend api configs...
^M 1 / 2 completed |^M 2 / 2 completed /Finished in 0.0s
        
Updating metric owners...
Scoped order and limit are ignored, it's forced to be batch order and batch size
^M 1 / 7 completed |^M 2 / 7 completed /^M 3 / 7 completed -^M 4 / 7 completed \^M 5 / 7 completed |^M 6 / 7 completed /^M 7 / 7 completed -Finished in 0.0s
        
Updating proxy rules owners...
^M 1 / 2 completed |^M 2 / 2 completed /Finished in 0.0s
```

It seems to be correct as they are 3 sections and we added 3 commands.

File: pkg/3scale/amp/component/system.go [hunk 1]
```diff
@@ -564,7 +564,7 @@ func (system *System) AppDeploymentConfig() *appsv1.DeploymentConfig {
 						ExecNewPod: &appsv1.ExecNewPodHook{
 							// TODO the MASTER_ACCESS_TOKEN reference should be probably set as an envvar that gathers its value from the system-seed secret
 							// but changing that probably has some implications during an upgrade process of the product
-							Command:       []string{"bash", "-c", "bundle exec rake boot openshift:deploy " + "MASTER_ACCESS_TOKEN" + "=\"" + system.Options.masterAccessToken + "\""},
+							Command:       []string{"bash", "-c", "bundle exec rake boot openshift:deploy services:create_backend_apis services:update_metric_owners proxy:update_proxy_rule_owners " + "MASTER_ACCESS_TOKEN" + "=\"" + system.Options.masterAccessToken + "\""},
```
Comment (eguzki) [inline pkg/3scale/amp/component/system.go]: `fmt.Sprintf` ?

**Comment by eguzki** [comment]
missing upgrade path

**Comment by miguelsorianod** [comment]
> missing upgrade path

I know. the PR title says WIP

File: pkg/3scale/amp/auto-generated-templates/amp/amp-eval-s3.yml [hunk 2]
```diff
@@ -1623,6 +1623,8 @@ objects:
             - bash
             - -c
             - bundle exec rake boot openshift:deploy MASTER_ACCESS_TOKEN="${MASTER_ACCESS_TOKEN}"
+              && bundle exec rake services:create_backend_apis services:update_metric_owners
```
Comment (miguelsorianod) [inline pkg/3scale/amp/auto-generated-templates/amp/amp-eval-s3.yml]: The new tasks have been added in a new bundle exec rake command due to it seems when used in the same bundle exec rake it failed. After commenting this with @guicassolato it seems that bundle exec rake loads everything in the same scope and the information that first tasks update is not visible to the others. That made the tasks fails. Changing it to separate tasks seems to have fixed the problem

File: pkg/controller/apimanager/upgrade_26_to_27.go [hunk 3]
```diff
@@ -12,7 +16,13 @@ type Upgrade26_to_27 struct {
 }
 
 func (u *Upgrade26_to_27) Upgrade() (reconcile.Result, error) {
-	res, err := u.upgradeAMPImageStreams()
+
+	res, err := u.upgradeSystemAppPreHookPodCommand()
```
Comment (miguelsorianod) [inline pkg/controller/apimanager/upgrade_26_to_27.go]: We execute the upgrade of the prehook pod command BEFORE the upgrade of the system image.
When the prehook pod command is executed, the prehook pod is NOT redeployed at that moment.

The upgradeSystemAppPreHookPodCommand tries to get the existing system-app DeploymentConfig and checks whether the existing command matches the desired one. When there's a difference the DC is updated and a requeue is performed. This means that at the next iteration of the reconciliation loop we ensure the change is visible when getting it again.

Finally, we upgrade AMP Images after updating the system-app prehook pod. That changes the ImageStream that triggers a DeploymentConfig redeploy, triggering the pre-hook pod execution.

**Comment by miguelsorianod** [comment]
pre-hook pod logs when performing an upgrade

```
msoriano@localhost:~/go/src/github.com/3scale/3scale-operator (fix-application-migration-upgrades)$ oc logs system-app-2-hook-pre
ActiveMerchant MODE set to 'production'
No valid API key has been set, notifications will not be sent
[Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 2.96.1 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
Migrating to CreateBackendApis (20190715144518)
== 20190715144518 CreateBackendApis: migrating ================================
-- create_table(:backend_apis)
   -> 0.0183s
-- create_table(:backend_api_configs)
   -> 0.0086s
-- execute("ALTER TABLE backend_apis CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;")
   -> 0.0131s
-- execute("ALTER TABLE backend_api_configs CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;")
   -> 0.0130s
== 20190715144518 CreateBackendApis: migrated (0.0534s) =======================
Migrating to AddOwnerToMetrics (20190716110000)
== 20190716110000 AddOwnerToMetrics: migrating ================================
-- add_column(:metrics, :owner_id, :integer, {:limit=>8})
   -> 0.0431s
-- add_column(:metrics, :owner_type, :string)
   -> 0.0562s
-- add_index(:metrics, [:owner_type, :owner_id], {})
   -> 0.0320s
-- add_index(:metrics, [:owner_type, :owner_id, :system_name], {:unique=>true})
   -> 0.0176s
== 20190716110000 AddOwnerToMetrics: migrated (0.1723s) =======================
Migrating to CreateTheBackendApisOfServices (20190716110520)
== 20190716110520 CreateTheBackendApisOfServices: migrating ===================
Nothing to do, this migration should not be executed
== 20190716110520 CreateTheBackendApisOfServices: migrated (0.0000s) ==========
Migrating to AddActsAsProductToService (20190731092006)
== 20190731092006 AddActsAsProductToService: migrating ========================
-- add_column(:services, :act_as_product, :boolean)
   -> 0.0466s
-- change_column_default(:services, :act_as_product, false)
   -> 0.0156s
== 20190731092006 AddActsAsProductToService: migrated (0.0624s) ===============
Migrating to BackfillAddActsAsProductToService (20190731092338)
== 20190731092338 BackfillAddActsAsProductToService: migrating ================
Nothing to do, this migration should not be executed
== 20190731092338 BackfillAddActsAsProductToService: migrated (0.0000s) =======
Migrating to AddOwnerToProxyRules (20190801143026)
== 20190801143026 AddOwnerToProxyRules: migrating =============================
-- add_column(:proxy_rules, :owner_id, :integer, {:limit=>8})
   -> 0.0404s
-- add_column(:proxy_rules, :owner_type, :string)
   -> 0.0427s
-- add_index(:proxy_rules, [:owner_type, :owner_id], {})
   -> 0.0646s
== 20190801143026 AddOwnerToProxyRules: migrated (0.1491s) ====================
Migrating to BackfillProxyRulesOwners (20190801143259)
== 20190801143259 BackfillProxyRulesOwners: migrating =========================
Nothing to do, this migration should not be executed
== 20190801143259 BackfillProxyRulesOwners: migrated (0.0000s) ================
Migrating to CreateTriggersBackendApis (20190802133303)
== 20190802133303 CreateTriggersBackendApis: migrating ========================
== 20190802133303 CreateTriggersBackendApis: migrated (0.1419s) ===============
Migrating to BackfillMetricOwners (20190805135829)
== 20190805135829 BackfillMetricOwners: migrating =============================
Nothing to do, this migration should not be executed
== 20190805135829 BackfillMetricOwners: migrated (0.0000s) ====================
Migrating to AddStateToBackendApis (20190904144157)
== 20190904144157 AddStateToBackendApis: migrating ============================
-- add_column(:backend_apis, :state, :string, {:default=>:published, :null=>false})
   -> 0.0118s
-- add_index(:backend_apis, :state, {})
   -> 0.0078s
== 20190904144157 AddStateToBackendApis: migrated (0.0198s) ===================
Migrating to CreateGatewayConfigurations (20190920123906)
== 20190920123906 CreateGatewayConfigurations: migrating ======================
-- create_table(:gateway_configurations)
   -> 0.0081s
-- add_index(:gateway_configurations, [:proxy_id], {:unique=>true})
   -> 0.0144s
== 20190920123906 CreateGatewayConfigurations: migrated (0.0227s) =============
Migrating to EnforceUniquenessAssociationServiceBackendApi (20190925133159)
== 20190925133159 EnforceUniquenessAssociationServiceBackendApi: migrating ====
-- add_index(:backend_api_configs, [:backend_api_id, :service_id], {:unique=>true})
   -> 0.0073s
== 20190925133159 EnforceUniquenessAssociationServiceBackendApi: migrated (0.0074s) 
Migrating to CreateProxyConfigAffectingChanges (20190925152107)
== 20190925152107 CreateProxyConfigAffectingChanges: migrating ================
-- create_table(:proxy_config_affecting_changes)
   -> 0.0144s
-- add_index(:proxy_config_affecting_changes, :proxy_id, {:unique=>true})
   -> 0.0415s
== 20190925152107 CreateProxyConfigAffectingChanges: migrated (0.0562s) =======
ActiveMerchant MODE set to 'production'
No valid API key has been set, notifications will not be sent
[Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Migrating proxies api_backend to backend api configs...
 2 / 2 completed /Enqueued BackendMetricWorker#6db3b48ccb01ebb96b3d9e12 with args: ["1", 6, "hits.1"]
Enqueued AuditedWorker#853708d68a6370788ade0a39 with args: [{"id"=>nil, "auditable_id"=>6, "auditable_type"=>"Metric", "user_id"=>nil, "user_type"=>nil, "username"=>nil, "action"=>"create", "version"=>1, "created_at"=>nil, "tenant_id"=>nil, "provider_id"=>nil, "kind"=>"Metric", "audited_changes"=>{"system_name"=>"hits.1", "description"=>"Number of API hits", "unit"=>"hit", "created_at"=>2019-10-14 14:31:47 UTC, "service_id"=>nil, "friendly_name"=>"Hits", "parent_id"=>nil, "tenant_id"=>nil, "owner_id"=>1, "owner_type"=>"BackendApi"}, "comment"=>nil, "associated_id"=>nil, "associated_type"=>nil, "remote_address"=>nil, "request_uuid"=>"c844271c-6130-4ad8-907a-786bdad1684a"}]
[EventBroker] notifying subscribers of ProxyConfigs::AffectingObjectChangedEvent ac5a4f05-c972-4ee5-b592-a0bff166a3b7
[ActiveJob] Enqueued ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper#e8e03114f3a302506fd1d32c with args: [{"job_class"=>"ProxyConfigAffectingChangeWorker", "job_id"=>"8ebd8aa6-f2ec-406d-ad94-27ae554d178a", "queue_name"=>"default", "arguments"=>["ac5a4f05-c972-4ee5-b592-a0bff166a3b7"], "locale"=>:en}]
[ActiveJob] Enqueued ProxyConfigAffectingChangeWorker (Job ID: 8ebd8aa6-f2ec-406d-ad94-27ae554d178a) to Sidekiq(default) with arguments: "ac5a4f05-c972-4ee5-b592-a0bff166a3b7"
Enqueued BackendMetricWorker#84810f06ce71314424fbab2a with args: ["1", 6, "hits.1"]
Enqueued BackendMetricWorker#dbf5481db3d88248d34c0724 with args: ["2", 7, "hits.2"]
Enqueued AuditedWorker#fb493d4c90140511b3d6e0ea with args: [{"id"=>nil, "auditable_id"=>7, "auditable_type"=>"Metric", "user_id"=>nil, "user_type"=>nil, "username"=>nil, "action"=>"create", "version"=>1, "created_at"=>nil, "tenant_id"=>nil, "provider_id"=>nil, "kind"=>"Metric", "audited_changes"=>{"system_name"=>"hits.2", "description"=>"Number of API hits", "unit"=>"hit", "created_at"=>2019-10-14 14:31:47 UTC, "service_id"=>nil, "friendly_name"=>"Hits", "parent_id"=>nil, "tenant_id"=>nil, "owner_id"=>2, "owner_type"=>"BackendApi"}, "comment"=>nil, "associated_id"=>nil, "associated_type"=>nil, "remote_address"=>nil, "request_uuid"=>"e61b6c45-e7ea-4f63-accf-78760a84149e"}]
[EventBroker] notifying subscribers of ProxyConfigs::AffectingObjectChangedEvent abe9900c-2d99-4e42-8f35-85a28f46a10a
[ActiveJob] Enqueued ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper#81e5e84fa756e764d3fb02fe with args: [{"job_class"=>"ProxyConfigAffectingChangeWorker", "job_id"=>"2fcbdc57-1532-4ab6-a33e-b01e67d8a565", "queue_name"=>"default", "arguments"=>["abe9900c-2d99-4e42-8f35-85a28f46a10a"], "locale"=>:en}]
[ActiveJob] Enqueued ProxyConfigAffectingChangeWorker (Job ID: 2fcbdc57-1532-4ab6-a33e-b01e67d8a565) to Sidekiq(default) with arguments: "abe9900c-2d99-4e42-8f35-85a28f46a10a"
Enqueued BackendMetricWorker#e51d1e3c4cb7c75c9bd01e36 with args: ["2", 7, "hits.2"]
Finished in 1.0s
	
Updating metric owners...
Scoped order and limit are ignored, it's forced to be batch order and batch size
 7 / 7 completed -Finished in 0.0s
	
Updating proxy rules owners...
 2 / 2 completed /Finished in 0.1s
	
msoriano@localhost:~/go/src/github.com/3scale/3scale-operator (fix-application-migration-upgrades)$ oc logs system-app-2-hook-pre
ActiveMerchant MODE set to 'production'
No valid API key has been set, notifications will not be sent
[Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 2.96.1 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
Migrating to CreateBackendApis (20190715144518)
== 20190715144518 CreateBackendApis: migrating ================================
-- create_table(:backend_apis)
   -> 0.0183s
-- create_table(:backend_api_configs)
   -> 0.0086s
-- execute("ALTER TABLE backend_apis CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;")
   -> 0.0131s
-- execute("ALTER TABLE backend_api_configs CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;")
   -> 0.0130s
== 20190715144518 CreateBackendApis: migrated (0.0534s) =======================
Migrating to AddOwnerToMetrics (20190716110000)
== 20190716110000 AddOwnerToMetrics: migrating ================================
-- add_column(:metrics, :owner_id, :integer, {:limit=>8})
   -> 0.0431s
-- add_column(:metrics, :owner_type, :string)
   -> 0.0562s
-- add_index(:metrics, [:owner_type, :owner_id], {})
   -> 0.0320s
-- add_index(:metrics, [:owner_type, :owner_id, :system_name], {:unique=>true})
   -> 0.0176s
== 20190716110000 AddOwnerToMetrics: migrated (0.1723s) =======================
Migrating to CreateTheBackendApisOfServices (20190716110520)
== 20190716110520 CreateTheBackendApisOfServices: migrating ===================
Nothing to do, this migration should not be executed
== 20190716110520 CreateTheBackendApisOfServices: migrated (0.0000s) ==========
Migrating to AddActsAsProductToService (20190731092006)
== 20190731092006 AddActsAsProductToService: migrating ========================
-- add_column(:services, :act_as_product, :boolean)
   -> 0.0466s
-- change_column_default(:services, :act_as_product, false)
   -> 0.0156s
== 20190731092006 AddActsAsProductToService: migrated (0.0624s) ===============
Migrating to BackfillAddActsAsProductToService (20190731092338)
== 20190731092338 BackfillAddActsAsProductToService: migrating ================
Nothing to do, this migration should not be executed
== 20190731092338 BackfillAddActsAsProductToService: migrated (0.0000s) =======
Migrating to AddOwnerToProxyRules (20190801143026)
== 20190801143026 AddOwnerToProxyRules: migrating =============================
-- add_column(:proxy_rules, :owner_id, :integer, {:limit=>8})
   -> 0.0404s
-- add_column(:proxy_rules, :owner_type, :string)
   -> 0.0427s
-- add_index(:proxy_rules, [:owner_type, :owner_id], {})
   -> 0.0646s
== 20190801143026 AddOwnerToProxyRules: migrated (0.1491s) ====================
Migrating to BackfillProxyRulesOwners (20190801143259)
== 20190801143259 BackfillProxyRulesOwners: migrating =========================
Nothing to do, this migration should not be executed
== 20190801143259 BackfillProxyRulesOwners: migrated (0.0000s) ================
Migrating to CreateTriggersBackendApis (20190802133303)
== 20190802133303 CreateTriggersBackendApis: migrating ========================
== 20190802133303 CreateTriggersBackendApis: migrated (0.1419s) ===============
Migrating to BackfillMetricOwners (20190805135829)
== 20190805135829 BackfillMetricOwners: migrating =============================
Nothing to do, this migration should not be executed
== 20190805135829 BackfillMetricOwners: migrated (0.0000s) ====================
Migrating to AddStateToBackendApis (20190904144157)
== 20190904144157 AddStateToBackendApis: migrating ============================
-- add_column(:backend_apis, :state, :string, {:default=>:published, :null=>false})
   -> 0.0118s
-- add_index(:backend_apis, :state, {})
   -> 0.0078s
== 20190904144157 AddStateToBackendApis: migrated (0.0198s) ===================
Migrating to CreateGatewayConfigurations (20190920123906)
== 20190920123906 CreateGatewayConfigurations: migrating ======================
-- create_table(:gateway_configurations)
   -> 0.0081s
-- add_index(:gateway_configurations, [:proxy_id], {:unique=>true})
   -> 0.0144s
== 20190920123906 CreateGatewayConfigurations: migrated (0.0227s) =============
Migrating to EnforceUniquenessAssociationServiceBackendApi (20190925133159)
== 20190925133159 EnforceUniquenessAssociationServiceBackendApi: migrating ====
-- add_index(:backend_api_configs, [:backend_api_id, :service_id], {:unique=>true})
   -> 0.0073s
== 20190925133159 EnforceUniquenessAssociationServiceBackendApi: migrated (0.0074s) 
Migrating to CreateProxyConfigAffectingChanges (20190925152107)
== 20190925152107 CreateProxyConfigAffectingChanges: migrating ================
-- create_table(:proxy_config_affecting_changes)
   -> 0.0144s
-- add_index(:proxy_config_affecting_changes, :proxy_id, {:unique=>true})
   -> 0.0415s
== 20190925152107 CreateProxyConfigAffectingChanges: migrated (0.0562s) =======
ActiveMerchant MODE set to 'production'
No valid API key has been set, notifications will not be sent
[Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Migrating proxies api_backend to backend api configs...
 2 / 2 completed /Enqueued BackendMetricWorker#6db3b48ccb01ebb96b3d9e12 with args: ["1", 6, "hits.1"]
Enqueued AuditedWorker#853708d68a6370788ade0a39 with args: [{"id"=>nil, "auditable_id"=>6, "auditable_type"=>"Metric", "user_id"=>nil, "user_type"=>nil, "username"=>nil, "action"=>"create", "version"=>1, "created_at"=>nil, "tenant_id"=>nil, "provider_id"=>nil, "kind"=>"Metric", "audited_changes"=>{"system_name"=>"hits.1", "description"=>"Number of API hits", "unit"=>"hit", "created_at"=>2019-10-14 14:31:47 UTC, "service_id"=>nil, "friendly_name"=>"Hits", "parent_id"=>nil, "tenant_id"=>nil, "owner_id"=>1, "owner_type"=>"BackendApi"}, "comment"=>nil, "associated_id"=>nil, "associated_type"=>nil, "remote_address"=>nil, "request_uuid"=>"c844271c-6130-4ad8-907a-786bdad1684a"}]
[EventBroker] notifying subscribers of ProxyConfigs::AffectingObjectChangedEvent ac5a4f05-c972-4ee5-b592-a0bff166a3b7
[ActiveJob] Enqueued ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper#e8e03114f3a302506fd1d32c with args: [{"job_class"=>"ProxyConfigAffectingChangeWorker", "job_id"=>"8ebd8aa6-f2ec-406d-ad94-27ae554d178a", "queue_name"=>"default", "arguments"=>["ac5a4f05-c972-4ee5-b592-a0bff166a3b7"], "locale"=>:en}]
[ActiveJob] Enqueued ProxyConfigAffectingChangeWorker (Job ID: 8ebd8aa6-f2ec-406d-ad94-27ae554d178a) to Sidekiq(default) with arguments: "ac5a4f05-c972-4ee5-b592-a0bff166a3b7"
Enqueued BackendMetricWorker#84810f06ce71314424fbab2a with args: ["1", 6, "hits.1"]
Enqueued BackendMetricWorker#dbf5481db3d88248d34c0724 with args: ["2", 7, "hits.2"]
Enqueued AuditedWorker#fb493d4c90140511b3d6e0ea with args: [{"id"=>nil, "auditable_id"=>7, "auditable_type"=>"Metric", "user_id"=>nil, "user_type"=>nil, "username"=>nil, "action"=>"create", "version"=>1, "created_at"=>nil, "tenant_id"=>nil, "provider_id"=>nil, "kind"=>"Metric", "audited_changes"=>{"system_name"=>"hits.2", "description"=>"Number of API hits", "unit"=>"hit", "created_at"=>2019-10-14 14:31:47 UTC, "service_id"=>nil, "friendly_name"=>"Hits", "parent_id"=>nil, "tenant_id"=>nil, "owner_id"=>2, "owner_type"=>"BackendApi"}, "comment"=>nil, "associated_id"=>nil, "associated_type"=>nil, "remote_address"=>nil, "request_uuid"=>"e61b6c45-e7ea-4f63-accf-78760a84149e"}]
[EventBroker] notifying subscribers of ProxyConfigs::AffectingObjectChangedEvent abe9900c-2d99-4e42-8f35-85a28f46a10a
[ActiveJob] Enqueued ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper#81e5e84fa756e764d3fb02fe with args: [{"job_class"=>"ProxyConfigAffectingChangeWorker", "job_id"=>"2fcbdc57-1532-4ab6-a33e-b01e67d8a565", "queue_name"=>"default", "arguments"=>["abe9900c-2d99-4e42-8f35-85a28f46a10a"], "locale"=>:en}]
[ActiveJob] Enqueued ProxyConfigAffectingChangeWorker (Job ID: 2fcbdc57-1532-4ab6-a33e-b01e67d8a565) to Sidekiq(default) with arguments: "abe9900c-2d99-4e42-8f35-85a28f46a10a"
Enqueued BackendMetricWorker#e51d1e3c4cb7c75c9bd01e36 with args: ["2", 7, "hits.2"]
Finished in 1.0s
	
Updating metric owners...
Scoped order and limit are ignored, it's forced to be batch order and batch size
 7 / 7 completed -Finished in 0.0s
	
Updating proxy rules owners...
 2 / 2 completed /Finished in 0.1s

```

They seem to be correct

**Comment by miguelsorianod** [comment]
@eguzki ready for review

File: pkg/controller/apimanager/upgrade_26_to_27.go (hunk 3 above)
Comment (eguzki) [reply pkg/controller/apimanager/upgrade_26_to_27.go]: `When the prehook pod command is updated, the prehook pod is NOT redeployed at that moment.` I do not understand. We have a trigger of type `configchange`. Then, I would expect that when you change the DC (pre-hook command is changing the DC), a new deployment would be triggered, hence prehook pod will be exected or redeployed.

**Comment by eguzki** [comment]
I think we should change imagestream tag used by the deployment config to a more specific release tag. 

Such generic tag like `amp-system:latest` may lead to inconsistencies. Pre-hook command works for a given release, 2.7 in this case, so having a floating tag more specific to the release, makes all the sense. This reasoning stands for any other container command line, that could be specific to a given 3scale release image series.

In addition, currently we are redeploying prehook pod with image 2.6, running a command that will fail because it is meant for 2.7 and then updating imagestream to upgrade new 2.7 image. Triggering second re-deployment, the one expected to succeed because the command is consistent with the image release version.

The idea would be something like:
```
image: amp-system:3scale2.7 
```
Which would be a floating tag for 3scale 2.7 based images.

Then it would be a single redeployment, with  reasonable assurance that prehook command will be associated to the right image.

File: pkg/controller/apimanager/upgrade_26_to_27.go (hunk 3 above)
Comment (miguelsorianod) [reply pkg/controller/apimanager/upgrade_26_to_27.go]: That's not how it works.
The config is changed but the pre-hook pod is only redeployed when the template of the DeploymentConfig changes.

**Comment by miguelsorianod** [comment]
>  In addition, currently we are redeploying prehook pod with image 2.6, running a command that will fail because it is meant for 2.7 and then updating imagestream to upgrade new 2.7 image. Triggering second re-deployment, the one expected to succeed because the command is consistent with the image release version.

This will not happen that's what I tried to explain in my comment. There should only be one redeployment.

Also, at this point of the release (after ER1) I would not change the way our ImageStreams are managed for all our components as this has potentially big implications for new and existing customers that we might not be aware of at this moment. We can think about it on the next release together with whether we should go to a fixed tag model or a rolling one.

File: pkg/3scale/amp/component/system.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/component/system.go]: :+1:

**Comment by miguelsorianod** [comment]
Also logs when forcing a redeploy of the pre-hook pod in an existing installation, without being an upgrade:

```
ActiveMerchant MODE set to 'production'
No valid API key has been set, notifications will not be sent
[Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
[core] non-native log levels verbose, notice, critical emulated using UNKNOWN severity
Backend Internal API version 2.96.1 status: ok
Connected to mysql2://root@system-mysql/system
Connected to redis://system-redis:6379/1
ActiveMerchant MODE set to 'production'
No valid API key has been set, notifications will not be sent
[Core] Using http://backend-listener:3000/internal/ as URL
OpenIdAuthentication.store is nil. Using in-memory store.
Migrating proxies api_backend to backend api configs...
 2 / 2 completed /Finished in 0.2s
	
Updating metric owners...
Scoped order and limit are ignored, it's forced to be batch order and batch size
 7 / 7 completed -Finished in 0.0s
	
Updating proxy rules owners...
 2 / 2 completed /Finished in 0.0s
```

File: pkg/controller/apimanager/upgrade_26_to_27.go (hunk 3 above)
Comment (eguzki) [reply pkg/controller/apimanager/upgrade_26_to_27.go]: right, according to doc: https://docs.openshift.com/container-platform/4.1/applications/deployments/managing-deployment-processes.html#deployments-triggers_deployment-operations
```
The ConfigChange trigger results in a new ReplicationController whenever configuration changes are detected in the Pod template of the DeploymentConfig.
```

Tried myself and changing only prehook command does to trigger creation of a new deployment process

**Comment by eguzki** [comment]
Having a single new deployment process, so far, there is no big benefit of changing imagestreams tagging schema which showed to be limited and could cause inconsistencies. Imagestreams schema should provide easy and flexible management of images, instead of limiting and setting constraints.

The impact of changing to a new imagestream schema does not make up for the little benefit at short term. 

I agree we should review tagging schema for new releases.

**Pull request merged**